kernel/src/expressions/scalars.rs

```diff
@@ -999,4 +1068,187 @@ mod tests {
         let null = Scalar::Null(DataType::INTEGER);
         assert!(!null.eq(&null));
     }
+
+    #[test]
+    fn test_hashmap_conversion() -> DeltaResult<()> {
+        // Create a simple HashMap with string keys and integer values
+        let mut map = HashMap::new();
+        map.insert("key1".to_string(), 42i32);
+        map.insert("key2".to_string(), 100i32);
+
+        // Convert HashMap to Scalar
+        let scalar = Scalar::try_from(map)?;
+
+        // Verify the scalar is of Map type
+        assert!(matches!(scalar, Scalar::Map(_)));
+
+        // Verify the data type
+        let expected_map_type = MapType::new(DataType::STRING, DataType::INTEGER, false);
+        assert_eq!(scalar.data_type(), DataType::from(expected_map_type));
+
+        // Extract the MapData and verify contents
+        if let Scalar::Map(map_data) = scalar {
+            let pairs = map_data.pairs();
+            assert_eq!(pairs.len(), 2);
+
+            // Check that both expected pairs are present
+            let has_key1 = pairs.iter().any(|(k, v)| {
+                matches!(k, Scalar::String(s) if s == "key1") && matches!(v, Scalar::Integer(42))
+            });
+            let has_key2 = pairs.iter().any(|(k, v)| {
+                matches!(k, Scalar::String(s) if s == "key2") && matches!(v, Scalar::Integer(100))
+            });
+
+            assert!(has_key1, "Missing key1 -> 42 pair");
+            assert!(has_key2, "Missing key2 -> 100 pair");
+        } else {
+            panic!("Expected Map scalar");
+        }
```

Comment:
nit: We can return early and remove the nesting
```suggestion
        let Scalar::Map(map_data) = scalar else {
            panic!("Expected Map scalar");
        }
        let pairs = map_data.pairs();
        assert_eq!(pairs.len(), 2);

        // Check that both expected pairs are present
        let has_key1 = pairs.iter().any(|(k, v)| {
            matches!(k, Scalar::String(s) if s == "key1") && matches!(v, Scalar::Integer(42))
        });
        let has_key2 = pairs.iter().any(|(k, v)| {
            matches!(k, Scalar::String(s) if s == "key2") && matches!(v, Scalar::Integer(100))
        });

        assert!(has_key1, "Missing key1 -> 42 pair");
        assert!(has_key2, "Missing key2 -> 100 pair");
```